pageserver/src/virtual_file.rs

```diff
@@ -601,23 +602,36 @@ impl VirtualFile {
         Ok(())
     }
 
-    pub async fn write_all(&mut self, mut buf: &[u8]) -> Result<(), Error> {
+    /// Writes `buf.slice(0..buf.bytes_init())`.
+    /// Returns the IoBuf that is underlying the BoundedBuf `buf`.
+    /// I.e., the returned value's `bytes_init()` method returns something different than the `bytes_init()` that was passed in.
+    /// It's quite brittle and easy to mis-use, so, we return the size in the Ok() variant.
+    pub async fn write_all<B: BoundedBuf>(&mut self, buf: B) -> (B::Buf, Result<usize, Error>) {
+        let nbytes = buf.bytes_init();
+        if nbytes == 0 {
+            return (Slice::into_inner(buf.slice_full()), Ok(0));
+        }
+        let mut buf = buf.slice(0..nbytes);
         while !buf.is_empty() {
```

Comment:
Does the call to `is_empty` go through the deref [here](https://github.com/neondatabase/tokio-epoll-uring/blob/868d2c42b5d54ca82fead6e8f2f233b69a540d3e/uring-common/src/buf/slice.rs#L84-L92) first? If not, I'm not sure this is correct.

In any case, I'd prefer `while buf.bytes_init() != 0`.

Reply:
Yes, it's from the deref.
Why do you think it's not correct?

Reply:
If it goes through the deref it's fine. I wasn't sure it does since `Vec::is_empty` also matches.

Reply:
> since Vec::is_empty also matches.

It doesn't for me?  Or am I missing some Deref impl to the underlying B::Buf?

Reply:
You're right. `Slice` is an actual struct so that hides the underlying storage.